specs/core/0_beacon-chain.md

```diff
@@ -1772,135 +1772,291 @@ For each `transfer` in `block.body.transfers`:
-  * Any [active validator](#dfn-active-validator) `index` from `previous_active_validator_indices` not in `previous_epoch_head_attester_indices` loses `base_reward(state, index)`.
-* Inclusion distance:
-  * Any [validator](#dfn-validator) `index` in `previous_epoch_attester_indices` gains `base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY // inclusion_distance(state, index)`
-
-Case 2: `epochs_since_finality > 4`:
-
-* Any [active validator](#dfn-active-validator) `index` from `previous_active_validator_indices` not in `previous_epoch_attester_indices`, loses `inactivity_penalty(state, index, epochs_since_finality)`.
-* Any [active validator](#dfn-active-validator) `index` from `previous_active_validator_indices` not in `previous_epoch_boundary_attester_indices`, loses `inactivity_penalty(state, index, epochs_since_finality)`.
-* Any [active validator](#dfn-active-validator) `index` from `previous_active_validator_indices` not in `previous_epoch_head_attester_indices`, loses `base_reward(state, index)`.
-* Any [active validator](#dfn-active-validator) `index` with `validator.slashed is True`, loses `2 * inactivity_penalty(state, index, epochs_since_finality) + base_reward(state, index)`.
-* Any [validator](#dfn-validator) `index` in `previous_epoch_attester_indices` loses `base_reward(state, index) - base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY // inclusion_distance(state, index)`
+```python
+def get_justification_and_finalization_deltas(state: BeaconState) -> Dict[Int -> Int]:
+    previous_active_validator_indices = get_active_validator_indices(state.validator_registry, previous_epoch)
+    epochs_since_finality = get_current_epoch(state) + 1 - state.finalized_epoch
+    active_validator_indices = [
+        i for i in range(len(state.validator_registry)) if
+        is_active_validator(state.validator_registry[i], get_current_epoch(state) - 1)
+    ]
+    # Initialize deltas that we are returning to zero
+    deltas = {index: 0 for index in range(len(state.validator_registry))}
+    # Some helper variables
+    boundary_attestations = get_previous_epoch_boundary_attestations(state)
+    boundary_attesting_balance = get_attesting_balance(boundary_attestations)
+    total_balance = get_previous_total_balance(state)
+    total_attesting_balance = get_attesting_balance(get_previous_epoch_attestations(state))
+    matching_head_attestations = [
+        a for a in get_previous_epoch_attestations(state) if
+        a.data.beacon_block_root == get_block_root(state, a.data.slot)
+    ]
+    matching_head_balance = get_attesting_balance(matching_head_attestations)
+    # Normal case
+    if epochs_since_finality <= 4:
+        for index in active_validator_indices:
+            # Expected FFG source
+            if index in get_attesting_indices(get_previous_epoch_attestations(state)):
```

Comment:
Reminder: `get_previous_epoch_attestations` undefined. It was `previous_epoch_attestations = [a for a in state.latest_attestations if previous_epoch == slot_to_epoch(a.data.slot)]`.